payments-core/src/main/java/com/stripe/android/paymentsheet/flowcontroller/DefaultFlowController.kt

```diff
@@ -41,103 +38,49 @@ import com.stripe.android.paymentsheet.model.SetupIntentClientSecret
 import com.stripe.android.paymentsheet.repositories.PaymentMethodsApiRepository
 import com.stripe.android.paymentsheet.repositories.StripeIntentRepository
 import com.stripe.android.view.AuthActivityStarter
+import dagger.Lazy
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.isActive
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
 import kotlinx.parcelize.Parcelize
-import javax.inject.Provider
+import javax.inject.Inject
+import javax.inject.Singleton
 
-internal class DefaultFlowController internal constructor(
-    private val appContext: Context,
-    viewModelStoreOwner: ViewModelStoreOwner,
+@Singleton
+internal class DefaultFlowController @Inject internal constructor(
+    // Properties provided through FlowControllerComponent.Builder
     private val lifecycleScope: CoroutineScope,
-    activityLauncherFactory: ActivityLauncherFactory,
     private val statusBarColor: () -> Int?,
     private val authHostSupplier: () -> AuthActivityStarter.Host,
     private val paymentOptionFactory: PaymentOptionFactory,
+    private val paymentOptionCallback: PaymentOptionCallback,
+    private val paymentResultCallback: PaymentSheetResultCallback,
+    // Properties provided through injection
     private val flowControllerInitializer: FlowControllerInitializer,
-    paymentControllerFactory: PaymentControllerFactory,
-    paymentFlowResultProcessorFactory:
-        (ClientSecret, Provider<String>, StripeApiRepository) ->
-        PaymentFlowResultProcessor<out StripeIntent, StripeIntentResult<StripeIntent>>,
     private val eventReporter: EventReporter,
     private val sessionId: SessionId,
-    defaultReturnUrl: DefaultReturnUrl,
-    private val paymentOptionCallback: PaymentOptionCallback,
-    private val paymentResultCallback: PaymentSheetResultCallback
+    private val paymentOptionActivityLauncher: ActivityResultLauncher<PaymentOptionContract.Args>,
+    private val googlePayActivityLauncher: ActivityResultLauncher<StripeGooglePayContract.Args>,
+    private val viewModel: FlowControllerViewModel,
+    private val stripeApiRepository: StripeApiRepository,
+    private val paymentController: PaymentController,
+    /**
+     * [PaymentConfiguration] is [Lazy] because the client might set publishableKey and
+     * stripeAccountId after creating a [DefaultFlowController].
+     */
+    private val lazyPaymentConfiguration: Lazy<PaymentConfiguration>,
```

Comment:
Note: here we use `Lazy` instead of `Provider` to mirror the logic before the change - we allow user to call `PaymentConfiguration.init` **once** after DefaultFlowController instantiation. Any `PaymentConfiguration.init` after that will not take effect.

Reply:
I agree with this approach, but we'd also need to make sure all other classes that use the publishable key behave the same way. Some of them like the `AnalyticsRequestFactory` which is a dependency of `EventReporter` receive a provider, so if the provider always reads from `PaymentConfiguration.getInstance(appContext)` they'll use the new key if it's set again.
Dagger can help us stop reading the key from different locations at different times by providing the same dependency to all of them. Something like a provider that reads the key from a lazy PaymentConfiguration should work. Once the lazy PaymentConfiguration is retrieved the first time it needs to be used we always get the key from that instance, and don't read from `PaymentConfiguration.getInstance(appContext)` anymore.
This is not new to this PR though, just something we can do better.

Reply:
Or, to simplify, we could just replace all dependencies on the publishable key with dependencies on a lazy `PaymentConfiguration`, which is a singleton. Once the first class reads from it we'll be using the same key everywhere while the app is alive.

Reply:
See comment below, I hadn't read the whole PR, you're already doing that 😀

Reply:
Yeah, we'll need to make sure there is only one actual `PaymentConfiguration.init` call during the during lifecycle of `FlowController`